packages/rpc/src/index.ts

```diff
@@ -62,89 +62,75 @@ export const Code = component('rpc' as Component, {
-  reqId: number = 0
-  requests = new Map<ReqId, Operation>()
-  stream: () => Promise<RequestStream>
-  handler: (tx: Tx) => void
-
-  constructor (stream: () => Promise<RequestStream>, handler: (tx: Tx) => void) {
-    this.stream = stream
-    this.handler = handler
-  }
+export abstract class RequestProcessor {
+  protected reqIndex: number = 0
+  protected requests = new Map<ReqId, DefferedPromise>()
+
+  protected abstract send (request: Request<any>): void
 
-  public process<T>(response: Response<T>): void {
+  protected process (response: Response<any>): void {
     if (response.id !== undefined) {
       const req = this.requests.get(response.id)
       if (req !== undefined) {
-        req.resolve(response)
+        if (response.error !== undefined) {
+          req.reject(response.error)
+        } else {
+          req.resolve(response.result)
+        }
       }
-    } else {
-      // From server transaction
-      this.handler((response.result as unknown) as Tx)
     }
   }
 
-  public onOpen (): void {
+  /**
+   * Reject all waited pending operations.
+   *
+   * This method is intended to be executed by protocol listening parts to
+   * cancal any pending requests to control UI is not hang.
+   *
+   * @param error - error to reject with.
+   */
+  protected reject (error: any): void {
     // We need to reply requests in case they are missed.
     for (const op of this.requests.entries()) {
-      op[1].request()
+      op[1].reject(error)
     }
+    this.requests.clear()
   }
 
-  public async request (method: string, ...params: any[]): Promise<any> {
-    const id = ++this.reqId
-    let resolve: <T> (value: T) => void = () => {}
-
-    const p = new Promise((r) => { resolve = r }) // eslint-disable-line
-
-    const op: Operation = {
-      id,
-      request: async () => {
-        (await this.stream()).send({ method, id, params })
-      },
-      resolve,
-      promise: p
-    }
-    this.requests.set(id, op)
+  protected async request (method: string, ...params: any[]): Promise<any> {
+    const id = ++this.reqIndex
+    const promise = new DefferedPromise()
+    this.requests.set(id, promise)
 
     // Send request
-    op.request()
+    this.send({ id, method, params })
 
     // Waiting to be complete.
-    const response = await op.promise
-    const err = response.error as string
-    if (err !== undefined) {
-      throw new Error(`Failed to process request ${err}`)
-    }
-    return response.result
-  }
-
-  async findAll<T extends Doc> (_class: Ref<Class<T>>, query: DocumentQuery<T>): Promise<T[]> {
-    const result = await this.request('findAll', _class, query)
-    return result as T[]
-  }
-
-  async tx (tx: Tx): Promise<void> {
-    await this.request('tx', tx)
-    // Process on server and return result.
-    this.handler(tx)
+    return await promise.promise.catch((reason) => {
+      throw new PlatformError(new Status(Severity.ERROR, Code.ConnectionError, { message: JSON.stringify(reason ?? '') }))
```

Comment:
Is it really "Connection Error" or something else can happen? I'd prefer to use `unknownErr()` here to pass potential PlatformError forward.

Reply:
Updated.